### DISCUSSION TITLE
FeedSDK problem

### DISCUSSION THREAD

**Discussion by sergisolanellas** [amazon-php/sp-api-sdk#365] [General]
Hi,

I'm trying to upload a feed in php, and I can create a "createFeedDocument" and upload the content of feed, but the problem is in the function CreateFeed. 

`
$factory = new Psr17Factory();
		$client = new Curl($factory);
		
		$configuration = Configuration::forIAMUser(
		    clientID, 
		    clientSECRET,
		    awsAccessKey, 
		    IAMsecurityKey
		);
		
		
		$logger = new Logger('name');
		//$logger->pushHandler(new StreamHandler(__DIR__ . '/sp-api-php.log', Logger::DEBUG));
		
		$sdk = SellingPartnerSDK::create($client, $factory, $factory, $configuration, $logger);
		
		$accessToken = $sdk->oAuth()->exchangeRefreshToken(refreshToken);
		
		echo '<PRE>';
		
		
		$specification = new CreateFeedDocumentSpecification;

		$feedDoc = $sdk->feeds()->createFeedDocument( $accessToken, 
														regionAPI,
														$specification->setContentType('text/xml; charset=utf-8') );
		
		
		$feedDocID = $feedDoc['feed_document_id'];
		$urlFeedUpload = $feedDoc['url'];
		
		
		//print_r($feedDoc);
			
			
		echo 'Feed id: '.$feedDocID.'<br>';
		echo 'URL Feed: '.$urlFeedUpload.'<br>';
		
		
		// Upload file
		
		$curl = curl_init();
		curl_setopt($curl, CURLOPT_URL, $urlFeedUpload);
		curl_setopt($curl, CURLOPT_UPLOAD, true);
		curl_setopt($curl, CURLOPT_PROTOCOLS, CURLPROTO_HTTPS);
		curl_setopt($curl, CURLOPT_HTTPHEADER, array('Content-Type: text/xml; charset=utf-8'));
		curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($curl, CURLOPT_BINARYTRANSFER, 1);
		curl_setopt($curl, CURLOPT_HEADER, false);
		curl_setopt($curl, CURLOPT_PUT, 1);
		curl_setopt($curl, CURLOPT_INFILE, fopen(dir_location.'feeds/autoprice/data_amazon_price_rules.txt', 'rb'));
		curl_setopt($curl, CURLOPT_INFILESIZE, filesize(dir_location.'feeds/autoprice/data_amazon_price_rules.txt'));
		#Only use below option on TEST environment if you have a self-signed certificate!!! On production this can cause security issues
		curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
		$response = curl_exec($curl);
		curl_close($curl);
		
				
		$specificationNewFeed = new CreateFeedSpecification([ 
							'feed_type' => '_MARS_AUTOMATE_PRICING_MULTI_MARKETPLACE_FEED_',
							'marketplace_ids' => marketplaceIDs,
							'input_feed_document_id' => $feedDocID							
										
		 ]);
		
                        
        $responseFeed = $sdk->feeds()->createFeed(
            $accessToken,
            regionAPI,
            $specificationNewFeed
        );
        
        print_r($responseFeed);`

The error appears when I createFeed:

Fatal error:  Uncaught AmazonPHP\SellingPartner\Exception\ApiException: [400] Error connecting to the API (https://sellingpartnerapi-eu.amazon.com/feeds/2021-06-30/feeds) in /vendor/amazon-php/sp-api-sdk/src/AmazonPHP/SellingPartner/Api/FeedsApi/FeedsSDK.php:291
Stack trace:
#0 (352): AmazonPHP\SellingPartner\Api\FeedsApi\FeedsSDK->createFeed()
#1 {main}
  thrown in vendor/amazon-php/sp-api-sdk/src/AmazonPHP/SellingPartner/Api/FeedsApi/FeedsSDK.php on line 291

Anyone can helpme? I don't know what is the problem.

Thanks

**Comment by tomaszhanc** [comment]
What is the exact response from Amazon? You pasted only the message from SDK, but it logs the whole response - there might be more information what's wrong.

`_MARS_AUTOMATE_PRICING_MULTI_MARKETPLACE_FEED_` is it valid feed type? There isn't something like that in the docs: https://developer-docs.amazon.com/sp-api/docs/feed-type-values

`marketplaceIDs` what do you have here?

**Reply by sergisolanellas** [reply]
Thanks to reply.

I'm using MWS version and now I'm migrating all code from this new version. 

The feed _MARS_AUTOMATE_PRICING_MULTI_MARKETPLACE_FEED_ I used to upload a file of automated pricing in MWS and it works, but now in the new API, it seems the feed is not allowed/correct.

Inside of marketplaceIDs I have an array of my Marketplace ID. This works because I use this in another functions.

If I read the log, now I see the feed _MARS_AUTOMATE_PRICING_MULTI_MARKETPLACE_FEED_ is invalid.

[2023-03-10T10:00:54.893841+01:00] name.DEBUG: Amazon Selling Partner API post request {"api":"Feeds","operation":"createFeed","response_correlation_id":"correlation_id_640af1c6c8b292.84770253","response_body":"{\n  \"errors\": [\n    {\n      \"code\": \"InvalidInput\",\n      \"message\": \"feed type __MARS_AUTOMATE_PRICING_MULTI_MARKETPLACE_FEED__ is invalid\",\n      \"details\": \"\"\n    }\n  ]\n}","response_headers":{"Date":["Fri, 10 Mar 2023 09:00:54 GMT"],"Content-Type":["application/json"],"Content-Length":["173"],"Connection":["keep-alive"],"x-amzn-RequestId":["60c76d91-f93f-4339-92cc-cfd89978accd"],"x-amzn-RateLimit-Limit":["0.0083"],"x-amz-apigw-id":["Bjq3HFI_joEF9eQ="],"X-Amzn-Trace-Id":["Root=1-640af1c6-4bad164f42bf177a5e0f948a"]},"response_status_code":400} []

There are any suggestion to upload the files of automated pricing using API ?

**Reply by tomaszhanc** [reply]
I don't know, but unfortunately, it looks like that FEED type is no longer supported by SP API.

**Reply by sergisolanellas** [reply]
If is not longer support, it will be a crap… It’s really annoying upload 4 o 5 times per day the file that modify the automatic prices. :(

**Reply by tomaszhanc** [reply]
Maybe, there is a different FEED that supports it. Have you checked it? I have never used the automatic prices feature, so I don't know how does it differ from regular price updates. To update a price on Amazon I use `POST_PRODUCT_PRICING_DATA` feed.

**Reply by sergisolanellas** [reply]
Hi!
Yes I used before POST_PRODUCT_PRICING_DATA and it’s not the same feed. The feed I refeer is a plain text file (with tabs), thats allow to update the maximum price and lower price inside a rule of auto pricing feature.
I discovered the name of this feed using the API to get all list of feeds. I search the mismatch date/time when uploaded the files, then appears in this list this feed. And then I tried to sent the file like a feed with this name and it worked. Since this day, I used this feed to upload the auto pricing file. I hope this kind of feed someday it will be again.

**Comment by devilica** [comment]
I found solution for this create product feed in Laravel with Amazon SP-API (https://github.com/amazon-php/sp-api-sdk):

```
composer require amazon-php/sp-api-sdk
composer require nyholm/psr7 -W


use AmazonPHP\SellingPartner\AccessToken;
use AmazonPHP\SellingPartner\Model\Feeds\CreateFeedDocumentSpecification;
use AmazonPHP\SellingPartner\Model\Feeds\CreateFeedSpecification;
use AmazonPHP\SellingPartner\Regions;
use AmazonPHP\SellingPartner\SellingPartnerSDK;
use App\Models\Amazon\AmazonAuth;
use Exception;
use GuzzleHttp\Handler\CurlFactory;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\App;
use Carbon\Carbon;
use Illuminate\Support\Facades\Storage;
use GuzzleHttp\Client;
use AmazonPHP\SellingPartner\Api\FeedsApi\FeedsSDK;
use AmazonPHP\SellingPartner\Configuration;
use AmazonPHP\SellingPartner\Exception\ApiException;
use AmazonPHP\SellingPartner\Exception\InvalidArgumentException;
use AmazonPHP\SellingPartner\HttpFactory;
use AmazonPHP\SellingPartner\HttpSignatureHeaders;
use AmazonPHP\SellingPartner\ObjectSerializer;
use Psr\Http\Client\ClientExceptionInterface;
use Psr\Http\Client\ClientInterface;
use Psr\Http\Message\RequestInterface;
use Psr\Log\LoggerInterface;
use AmazonPHP\SellingPartner\OAuth;
use Buzz\Client\Curl;
use Nyholm\Psr7\Factory\Psr17Factory;
use Psr\Log\NullLogger;


public function getAccessToken($channelId)
  {

    $amazonAuth = AmazonAuth::where('channel_id', $channelId)->first();  //my local database channel
    $isExpired = Carbon::parse($amazonAuth->expires_at)->isPast();
    $token = null;

    if ($isExpired) {

      $data = [
        'grant_type' => 'refresh_token',
        'refresh_token' => $amazonAuth->refresh_token,
        'client_id' => TenantService::getLwaClientId(),
        'client_secret' => TenantService::getLwaSecret()
      ];

      $response = Http::asForm()->post(
        'https://api.amazon.com/auth/o2/token',
        $data
      );


      if ($response->successful()) {
        $body = $response->json();
        $amazonAuth->access_token = $body['access_token'];
        $amazonAuth->expires_at = now()->addSeconds($body['expires_in'])->toDateTimeString();
        $amazonAuth->save();

        $token = $body['access_token'];
      } else {
        Log::channel('amazon')->info('Amazon access token refresh failed for amazon auth id: ' . $amazonAuth->id);
       
      }

    } else {
      $token = $amazonAuth->access_token;
    }

    if ($token) {
      $accessToken = new AccessToken(
        $token,
        $amazonAuth->refresh_token,
        'refresh_token',
        (int) $amazonAuth->expires_in,
        'refresh_token'
      );

      return $accessToken;
    }


  }

public function createFeedTest()
  {

    $client = new Client();
    $factory = new Psr17Factory();
    // $httpFactory = new HttpFactory($factory, $factory);
    $region = 'eu';
    $accessToken = $this->getAccessToken(150);
    $logger = new NullLogger();
    $configuration = Configuration::forIAMUser(
      $getLwaClientId,
      $getLwaSecret,
      $getAwsAccessKey,
      $getAwsSecretKey,
    );
    $sdk = SellingPartnerSDK::create($client, $factory, $factory, $configuration, $logger);
    $region = Regions::EUROPE;

    $specification = new CreateFeedDocumentSpecification;
    $feedDoc = $sdk->feeds()->createFeedDocument(
      $accessToken,
      $region,
      $specification->setContentType('text/xml; charset=utf-8')
    );
    $feedDocID = $feedDoc['feed_document_id'];
    $urlFeedUpload = $feedDoc['url'];

    $fileContent = '<?xml version="1.0" encoding="utf-8" ?>
    <AmazonEnvelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              xsi:noNamespaceSchemaLocation="amzn-envelope.xsd">
            <Header>
              <DocumentVersion>1.01</DocumentVersion>
              <MerchantIdentifier>MYMERCHANTTOKEN</MerchantIdentifier>
            </Header>
            <MessageType>Product</MessageType>
            <PurgeAndReplace>false</PurgeAndReplace>
            <Message>
              <MessageID>1</MessageID>
              <OperationType>Update</OperationType>
              <Product>
                <SKU>56789</SKU>
                <StandardProductID>
                  <Type>ASIN</Type>
                  <Value>B0EXAMPLEG</Value>
                </StandardProductID>
                <ProductTaxCode>A_GEN_NOTAX</ProductTaxCode>
                <DescriptionData>
                  <Title>Example Product Title</Title>
                  <Brand>Example Product Brand</Brand>
                  <Description>This is an example product description.</Description>
                  <BulletPoint>Example Bullet Point 1</BulletPoint>
                  <BulletPoint>Example Bullet Point 2</BulletPoint>
                  <MSRP currency="USD">25.19</MSRP>
                  <Manufacturer>Example Product Manufacturer</Manufacturer>
                  <ItemType>example-item-type</ItemType>
                  <CountryOfOrigin>DE</CountryOfOrigin>
                  <UnitCount>1</UnitCount>
                  <PPUCountType>stück</PPUCountType>
                  <IsExpirationDatedProduct>false</IsExpirationDatedProduct>
                </DescriptionData>
                <ProductData>
                  <Health>
                    <ProductType>
                      <HealthMisc>
                        <Ingredients>Example Ingredients</Ingredients>
                        <Directions>Example Directions</Directions>
                      </HealthMisc>
                    </ProductType>
                  </Health>
                </ProductData>
                <IsHeatSensitive>false</IsHeatSensitive>

              </Product>
            </Message>
          </AmazonEnvelope>';

    dump($feedDoc);
    $curl = curl_init();
    curl_setopt($curl, CURLOPT_URL, $urlFeedUpload);
    curl_setopt($curl, CURLOPT_UPLOAD, true);
    curl_setopt($curl, CURLOPT_PROTOCOLS, CURLPROTO_HTTPS);
    curl_setopt($curl, CURLOPT_HTTPHEADER, array('Content-Type: text/xml; charset=utf-8'));
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($curl, CURLOPT_BINARYTRANSFER, 1);
    curl_setopt($curl, CURLOPT_HEADER, false);
    curl_setopt($curl, CURLOPT_PUT, 1);
    curl_setopt($curl, CURLOPT_INFILE, fopen('data://text/plain,' . $fileContent, 'r'));
    curl_setopt($curl, CURLOPT_INFILESIZE, strlen($fileContent));
    #Only use below option on TEST environment if you have a self-signed certificate!!! On production this can cause security issues
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
    $response = curl_exec($curl);
    dump($response);
    curl_close($curl);
    $specificationNewFeed = new CreateFeedSpecification([
      'feed_type' => 'POST_PRODUCT_DATA',
      'marketplace_ids' => ['A1PA6795UKMFR9'],
      'input_feed_document_id' => $feedDocID

    ]);


    $responseFeed = $sdk->feeds()->createFeed(
      $accessToken,
      $region,
      $specificationNewFeed
    );
    dd($responseFeed);
  }
```

**Reply by gala06195** [reply]
This saved me!!
